moshi/src/main/java/com/squareup/moshi/Moshi.java

```diff
@@ -133,11 +156,19 @@
           synchronized (adapterCache) {
             adapterCache.put(cacheKey, result);
           }
+          // Remove the type or field name only when we succeed in creating the adapter,
+          // so we have the full stack  at the top level.
+          typesAndFieldNames.remove(typesAndFieldNames.size() - 1);
           return result;
         }
       }
+    } catch (RuntimeException e) {
+      if (lastIndex == 0) { // Rewrite the exception at the top level.
+        e = errorWithFields(typesAndFieldNames, e);
+      }
+      throw e;
     } finally {
-      deferredAdapters.remove(deferredAdapters.size() - 1);
+      deferredAdapters.remove(lastIndex);
```

Comment:
it’s weird that we remove from `deferredAdapters` here, but from `typesAndFieldNames` in the loop. Should be the same place.

Reply:
yeah, can do!